In [99]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os


path = os.getcwd() + "/Titanic/data"
data = pd.read_csv(path + '/train.csv',header=0)
test = pd.read_csv(path + '/test.csv', header=0)

#### Первичный анализ данных

In [100]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [101]:
test.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
PassengerId    418 non-null int64
Pclass         418 non-null int64
Name           418 non-null object
Sex            418 non-null object
Age            332 non-null float64
SibSp          418 non-null int64
Parch          418 non-null int64
Ticket         418 non-null object
Fare           417 non-null float64
Cabin          91 non-null object
Embarked       418 non-null object
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [102]:
print(data.columns)
data.head(5)


Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [103]:
data.describe()


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

#### Исключаем из рассмотрения второстепенные и лишние данные  

In [104]:
data = data.drop(['Name', 'Ticket', 'Cabin'],axis=1)


#### Исключаем строки с пропущенными данными (NaN)

In [105]:
data = data.dropna()


#### Конвертация категориальных признаков 'Pclass', 'Embarked', 'Sex'

In [106]:
print(data['Pclass'].value_counts(),'\n')
print(data['Embarked'].value_counts(),'\n')
print(data['Sex'].value_counts(),'\n')


3    355
1    184
2    173
Name: Pclass, dtype: int64 

S    554
C    130
Q     28
Name: Embarked, dtype: int64 

male      453
female    259
Name: Sex, dtype: int64 



In [107]:
Pclass_ext = pd.get_dummies(data['Pclass'], prefix='Pclass')
Embarked_ext = pd.get_dummies(data['Embarked'], prefix='Embarked')
Sex_ext = pd.get_dummies(data['Sex'], prefix='Sex')
Embarked_ext.head()


Embarked_C  Embarked_Q  Embarked_S
0           0           0           1
1           1           0           0
2           0           0           1
3           0           0           1
4           0           0           1

In [108]:
data = pd.concat([data, Pclass_ext], axis=1)
data = pd.concat([data, Embarked_ext], axis=1)
data = pd.concat([data, Sex_ext], axis=1)
data.head(3)


PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0            1         0       3    male  22.0      1      0   7.2500   
1            2         1       1  female  38.0      1      0  71.2833   
2            3         1       3  female  26.0      0      0   7.9250   

  Embarked  Pclass_1  Pclass_2  Pclass_3  Embarked_C  Embarked_Q  Embarked_S  \
0        S         0         0         1           0           0           1   
1        C         1         0         0           1           0           0   
2        S         0         0         1           0           0           1   

   Sex_female  Sex_male  
0           0         1  
1           1         0  
2           1         0

In [109]:
data = data.drop(['Pclass', 'Embarked', 'Sex', 'PassengerId'], axis=1)
data.head(3)


Survived   Age  SibSp  Parch     Fare  Pclass_1  Pclass_2  Pclass_3  \
0         0  22.0      1      0   7.2500         0         0         1   
1         1  38.0      1      0  71.2833         1         0         0   
2         1  26.0      0      0   7.9250         0         0         1   

   Embarked_C  Embarked_Q  Embarked_S  Sex_female  Sex_male  
0           0           0           1           0         1  
1           1           0           0           1         0  
2           0           0           1           1         0

#### Смотрим, что получилось:

In [110]:
data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 712 entries, 0 to 890
Data columns (total 13 columns):
Survived      712 non-null int64
Age           712 non-null float64
SibSp         712 non-null int64
Parch         712 non-null int64
Fare          712 non-null float64
Pclass_1      712 non-null uint8
Pclass_2      712 non-null uint8
Pclass_3      712 non-null uint8
Embarked_C    712 non-null uint8
Embarked_Q    712 non-null uint8
Embarked_S    712 non-null uint8
Sex_female    712 non-null uint8
Sex_male      712 non-null uint8
dtypes: float64(2), int64(3), uint8(8)
memory usage: 38.9 KB


#### Разделение данных на обучающий и проверочный набор

In [111]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import accuracy_score

Y = data['Survived'].values
X = data.drop(['Survived'], axis=1).values

X_train, X_test, Y_Train, Y_test = train_test_split(X,Y,test_size=0.3,random_state=33)

#### Используем модели Линейной и Логистической регрессии
1. Создание модели
2. Обучение модели

In [112]:
model_log_regr = LogisticRegression()
model_lin_regr = LinearRegression()

model_log_regr.fit(X_train, Y_Train)
model_lin_regr.fit(X_train, Y_Train)


D:\Applications\Anaconda\envs\ml\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

#### Оценка точности моделей на проверочном наборе данных

In [114]:
predictions_log_regr = model_log_regr.predict(X_test)
predictions_lin_regr = model_lin_regr.predict(X_test).round()

accuracy_log_regr = accuracy_score(Y_test, predictions_log_regr).round(2)
accuracy_lin_regr = accuracy_score(Y_test, predictions_lin_regr).round(2)
print( f' -- Model accuracy--\nLogistic Regression: {accuracy_log_regr}\nLinear Regression: {accuracy_lin_regr}')

 -- Model accuracy--
Logistic Regression: 0.77
Linear Regression: 0.78


### Выводы
1. Использовать методы машинного обучения не сложно
2. Большая часть работы была посвящена подготовке данных
3. Точность моделей менее 78%. Это немного, однако, эта точность получена даром, без усилий

#### На самом деле
1. Нужно проводить более тщательный анализ исходных данных, выявлять закономерности и особенности.
2. Точность использованных моделей можно повысить при полном и правильном использовании данных.
3. Есть другие модели, которые лучше подходят для решения этой задачи.


